hatchery/src/snapshot_bag.rs

```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::error::Error;
+use crate::snapshot::Snapshot;
+use crate::snapshot::{MemoryPath, ModuleSnapshotId};
+use crate::Error::SnapshotError;
+
+#[derive(Debug)]
+pub struct SnapshotBag {
+    // first snapshot is always uncompressed
+    ids: Vec<ModuleSnapshotId>,
+    // we keep top uncompressed snapshot to make save snapshot efficient
+    top: ModuleSnapshotId,
+}
+
+impl SnapshotBag {
+    pub fn new() -> Self {
+        Self {
+            ids: Vec::new(),
+            top: ModuleSnapshotId::random(),
+        }
+    }
+    pub fn save_snapshot(
+        &mut self,
+        snapshot: &Snapshot,
+        memory_path: &MemoryPath,
+    ) -> Result<usize, Error> {
+        snapshot.capture(memory_path)?;
+        self.ids.push(snapshot.id());
+        if self.ids.len() == 1 {
+            // top is an uncompressed version of most recent snapshot
+            Snapshot::from_id(self.top, memory_path)?.capture(memory_path)?;
+            Ok(0)
+        } else {
+            let from_id =
+                |snapshot_id| Snapshot::from_id(snapshot_id, memory_path);
+            let top_snapshot = from_id(self.top)?;
+            let accu_snapshot = from_id(ModuleSnapshotId::random())?;
+            accu_snapshot.capture(snapshot)?;
+            // accu and snapshot are both uncompressed
+            // compressing snapshot against the top
+            snapshot.capture_diff(&top_snapshot, memory_path)?;
+            // snapshot is compressed but accu keeps the uncompressed copy
+            // top is an uncompressed version of most recent snapshot
+            top_snapshot.capture(&accu_snapshot)?;
+            Ok(self.ids.len() - 1)
+        }
+    }
+    pub fn restore_snapshot(
+        &self,
+        snapshot_index: usize,
```

Comment:
The consumers should not have to know about snapshot indexes. (using snapshot_index in a pub method)